experimental/segmenter/src/provider/lstm.rs

```diff
@@ -0,0 +1,362 @@
+        } else {
+            Ok(Self { dims, data })
+        }
+    }
+
+    #[doc(hidden)] // databake
+    pub const fn from_parts_unchecked(dims: [u16; D], data: ZeroVec<'data, f32>) -> Self {
+        Self { dims, data }
+    }
+
+    pub(crate) fn as_matrix_zero(&self) -> MatrixZero<D> {
+        MatrixZero::from_parts_unchecked(&self.data, self.dims.map(|x| x as usize))
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de: 'data, 'data, const D: usize> serde::Deserialize<'de> for LstmMatrix<'data, D> {
+    fn deserialize<S>(deserializer: S) -> Result<Self, S::Error>
+    where
+        S: serde::de::Deserializer<'de>,
+    {
+        #[derive(serde::Deserialize)]
+        struct Raw<'data> {
+            // no const generic support in serde
+            dims: alloc::borrow::Cow<'data, [u16]>,
+            #[serde(borrow)]
+            data: ZeroVec<'data, f32>,
+        }
+
+        let raw = Raw::deserialize(deserializer)?;
+
+        use core::ops::Deref;
+        use serde::de::Error;
+        Self::from_parts(
+            raw.dims
+                .deref()
+                .try_into()
+                .map_err(|_| S::Error::custom("Dimension mismatch"))?,
+            raw.data,
+        )
+        .map_err(|_| S::Error::custom("Dimension mismatch"))
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl<const D: usize> serde::Serialize for LstmMatrix<'_, D> {
+    fn serialize<S: serde::ser::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::SerializeStruct;
+        let mut state = serializer.serialize_struct("LstmMatrix", 2)?;
+        state.serialize_field("dims", self.dims.as_slice())?;
+        state.serialize_field("data", &self.data)?;
+        state.end()
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl<const D: usize> databake::Bake for LstmMatrix<'_, D> {
+    fn bake(&self, env: &databake::CrateEnv) -> databake::TokenStream {
+        let dims = self.dims.bake(env);
+        let data = self.data.bake(env);
+        databake::quote! {
+            icu_segmenter::provider::LstmMatrix::from_parts_unchecked(#dims, #data)
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Clone, Copy)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize,databake::Bake),
+    databake(path = icu_segmenter::provider),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+/// The type of LSTM model
+pub enum ModelType {
+    /// A model working on code points
+    Codepoints,
+    /// A model working on grapheme clusters
+    GraphemeClusters,
+}
+
+/// The struct that stores a LSTM model.
+#[icu_provider::data_struct(LstmDataV1Marker = "segmenter/lstm@1")]
+#[derive(PartialEq, Debug, Clone)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[yoke(prove_covariance_manually)]
+pub struct LstmDataV1<'data> {
+    /// Type of the model
+    pub(crate) model: ModelType,
+    /// The grapheme cluster dictionary used to train the model
```

Comment:
Nit: These fields need to encode the size invariants. It would be good to state them in the comments for each field.